### ISSUE TITLE
Ablate adaptive integrate-and-fire gating for sparse Player events

### ISSUE THREAD

**Issue by rhythmatician** [rhythmatician/FractalSync#152]
Part of #77

Coordinates with #104 (causal Player trunk/state architecture), #102 (SongFormer boundary/readiness distillation), #131 (Phase-1 representation pretraining), #107 (Controls v2 semantics), and #111 (eyes-on diagnostics/evaluation).

- **Status:** research
- **Primary question:** whether explicit integrate-threshold-reset dynamics improve timing and hysteresis for genuinely sparse/event-like Player actions
- **Important scope boundary:** this ticket does **not** propose a spiking Player, spike-encoded AudioActivation, or SongFormer-triggered runtime actions

## Motivation

The Player is intended to be a small persistent temporal policy rather than a frame-by-frame mapper:

```text
a_t = pi_theta(O_{<=t}, h_{t-1})
```

Most Controls v2 are persistent continuous quantities: steering, throttle, brake, grip/drift, Julia zoom/rotation deltas, and continuous ColorIntent evolution. Those should remain ordinary continuous outputs.

A smaller subset of behavior is genuinely event-like: bounded impulses/taps, a sparse kick-like gesture, discrete mode changes if retained, or another punctuated action that should happen once after enough musical evidence has accumulated and then should not immediately repeat.

A Leaky Integrate-and-Fire (LIF) or Adaptive LIF (ALIF) unit provides an explicit mechanism for that narrow temporal decision problem:

```text
weak evidence accumulates
        -> leak removes stale evidence
        -> threshold creates commitment
        -> reset/refractory state suppresses immediate retriggering
```

This may be preferable to requiring a generic recurrent trunk and unconstrained output head to rediscover temporal accumulation, hysteresis, thresholding, and cooldown behavior statistically.

The hypothesis is narrow and falsifiable:

> **Given an otherwise useful causal Player hidden state `h_t`, does explicit LIF/ALIF gating improve sparse event timing and reduce chatter compared with ordinary learned heads and simple hysteresis/cooldown baselines?**

## Architectural separation

This ticket is orthogonal to #104.

#104 chooses the minimal causal temporal trunk and defines persistent hidden-state semantics. This ticket assumes a trunk has produced a causal state:

```text
PlayerObservation_t
        |
        v
causal temporal trunk
        |
        v
       h_t
       |
       +------------------------> continuous Controls heads
       |                           steer / throttle / brake / grip
       |                           Julia zoom / rotation / palette evolution
       |
       `-> event evidence head
                |
                v
             LIF / ALIF
                |
             fire gate
                |
                v
          sparse event payload
```

Do not conflate trunk selection with event-gating selection. The experiment should be possible with the same frozen or matched trunk across variants.

## Candidate LIF formulation

A discrete-time baseline compatible with canonical Player ticks may be written as:

```text
I_t = w^T h_t + b
V_{t+1} = lambda_V * V_t + I_t
s_t = 1[V_{t+1} >= theta]
V_{t+1} <- V_reset        when s_t = 1
```

where:

- `V_t` is accumulated event evidence;
- `lambda_V` controls leak / evidence lifetime;
- `theta` is a firing threshold;
- `s_t` is a sparse binary event gate.

Equivalent continuous-time notation is acceptable for analysis:

```text
tau dV/dt = -V + I(t)
V >= theta -> fire; reset
```

The implementation must use explicit canonical-tick semantics and must not depend on render cadence.

## Adaptive LIF candidate

ALIF is the stronger hypothesis because adaptive threshold/refractory behavior maps directly onto:

> do something significant, then do not immediately do it again unless substantially new evidence arrives.

A candidate form is:

```text
I_t = w^T h_t + b
V_{t+1} = lambda_V * V_t + I_t
s_t = 1[V_{t+1} >= theta_t]

a_{t+1} = lambda_a * a_t + s_t
theta_t = theta_0 + beta_a * a_t
```

with voltage reset or subtractive reset after firing.

The exact state equations are research parameters, not a frozen runtime contract. If ALIF is retained, document/reset/version all persistent event-state semantics explicitly.

## Gate timing; do not spike-code the payload

The event spike should primarily answer **when to commit**, not encode the full analog action.

Prefer a factorization such as:

```text
payload_t = f_payload(h_t)
gate_t    = ALIF(f_event(h_t), event_state_t)
J_t       = gate_t * payload_t
```

For an impulse example:

```text
J_t = s_t * j_t
```

where:

- `s_t` decides whether the event occurs;
- `j_t` remains an ordinary bounded analog payload carrying direction/magnitude under Controls v2.

This avoids using spike rate/timing to represent quantities that are naturally continuous.

## Relationship to SongFormer readiness

#102/#131 may teach the causal shared state to represent transition readiness from training-only future-aware supervision.

This ticket may test whether such a learned readiness representation is useful input to an event integrator:

```text
music evidence
    -> causal hidden state h_t
    -> latent readiness/event evidence
    -> integrate
    -> threshold
    -> commit
```

However:

- SongFormer outputs remain target/loss-side only;
- no teacher probability or future-aware readiness value may directly charge a runtime LIF unit;
- the event bank consumes only causal Student state available at runtime;
- changing future audio may change offline targets but may not change the already-constructed runtime event state at `t`.

## Candidate event classes

Start only with actions that are genuinely sparse enough to justify event gating.

Initial candidates may include:

- bounded generalized impulse/tap under Controls v2;
- retained kick-like special action if it survives the Controls/debug surface;
- a discrete Julia/palette mode transition if such a discrete action remains in the product contract;
- another explicitly identified one-shot gesture with a measurable chatter/retrigger problem.

Do not create artificial event classes merely to exercise the mechanism.

## Continuous controls stay continuous

Do **not** route ordinary persistent controls through spikes by default:

```text
steer
throttle
brake
grip / drift
continuous Julia zoom delta/rate
continuous Julia rotation delta/rate
continuous palette / ColorIntent evolution
```

If one of these later needs hysteresis, test an appropriate continuous-state mechanism rather than assuming spike coding.

## Do not spike-encode AudioActivation

AudioActivation deliberately preserves continuous causal information such as cycle amplitude/strength, phase, continuous frequency, confidence, intensity, spectral distribution, and texture evidence.

Do not convert raw audio or AudioActivation into spikes as part of this experiment.

The event bank receives learned causal hidden state `h_t`, not raw waveform/spectrogram/CycleBank spike trains.

## Required baselines

An ALIF result is uninterpretable unless compared against simpler ways to get the same behavioral effect.

At minimum compare under matched data/trunk/Controls conditions:

| Variant | Event commitment mechanism |
|---|---|
| **A. Plain head** | ordinary learned event logit/value from `h_t`; no explicit temporal accumulator |
| **B. Hysteresis/cooldown** | ordinary head plus minimal explicit threshold/cooldown or Schmitt-like hysteresis |
| **C. LIF** | learned evidence -> leaky accumulator -> fixed threshold/reset |
| **D. ALIF** | learned evidence -> leaky accumulator -> adaptive threshold/refractory state |

If differentiable training requires surrogate gradients for C/D, document the estimator and compare training stability. A simpler detached/event-layer approach is also valid if the trunk/event-evidence head can be trained without making the discrete mechanism opaque.

The simple hysteresis/cooldown baseline is mandatory. Otherwise an ALIF win may only demonstrate that refractory behavior helps, not that adaptive integrate-and-fire dynamics are useful.

## Parameters and learning policy

Candidate parameters include:

```text
leak / tau_V
base threshold theta_0
reset rule / reset value
adaptation leak / tau_a
adaptation gain beta_a
optional minimum refractory interval
```

Compare at least one simple hand-set/stable parameterization against a learned-parameter version before accepting extra learnable state.

Constrain learned parameters so training cannot silently produce unstable or meaningless dynamics, e.g. bounded leaks/positive time constants and documented threshold ranges.

## Reset and persistent-state semantics

Event state is persistent causal state and must follow the same operational rigor as #104 hidden state.

Define behavior for:

- stream start;
- song/excerpt reset;
- training episode/reset under #110;
- truncated BPTT boundaries;
- checkpoint save/load;
- browser initialization;
- model hot reload where applicable;
- canonical analysis ticks continuing while render frames are late/dropped.

A gradient truncation boundary must not silently reset forward LIF/ALIF state unless that reset is an explicit experiment.

## Measurements

Primary evidence is downstream Player behavior, not membrane-voltage aesthetics or spike sparsity by itself.

Measure at least:

### Event chatter / redundancy

- events per minute;
- duplicate events within short windows;
- event burst-length distribution;
- rapid fire-reset-fire patterns;
- unnecessary impulses/mode changes during otherwise stable passages.

### Timing / commitment quality

Against held-out structural evidence and human judgment:

- event timing relative to actual musical transitions;
- premature commitments;
- late commitments;
- useful preparation followed by restraint when a suspected transition does not materialize;
- ability to accumulate repeated weak evidence rather than react to one noisy frame.

### Refractory / hysteresis behavior

- time to retrigger after a significant event;
- whether genuinely new evidence can override adaptation when appropriate;
- whether adaptation suppresses legitimate rapid musical repetition.

### Primary visual task

- downstream reward where valid;
- Shore-crossing/event timing where the event is relevant;
- Julia presentation coherence where a discrete presentation gesture is tested;
- eyes-on held-out judgment under #111;
- no degradation of ordinary continuous motion/presentation.

### Runtime/training cost

- additional persistent-state width;
- parameter count;
- CPU/browser inference cost;
- ONNX/export complexity;
- training stability / surrogate-gradient sensitivity if used.

## Controlled probes

Before full policy training, use synthetic event-evidence traces to verify the mechanism itself:

- isolated impulse;
- repeated subthreshold pulses;
- slow ramp;
- noisy near-threshold evidence;
- two legitimate events separated by varying intervals;
- stale evidence followed by silence;
- sustained strong evidence;
- alternating weak/strong cues.

Verify expected leak, threshold, reset, and adaptation behavior deterministically.

These probes test mechanism semantics only; they do not establish musical usefulness.

## Ablation design

Keep comparisons attributable.

Prefer:

```text
same PlayerObservation
same trunk architecture/checkpoint initialization policy
same Controls v2 contract
same training/evaluation songs and reset distributions
same event payload head capacity
only commitment mechanism changes
```

If #104 has not selected the final trunk yet, run the event-gating comparison on one stable recurrent baseline first. Repeat only the winning/minimal comparison on the selected v1 trunk if needed.

Do not cross-product every #104 trunk candidate with every LIF variant unless evidence shows an interaction worth investigating.

## Success criterion

Retain LIF/ALIF only if it provides a behavioral advantage that simpler mechanisms do not.

A positive result should look like some combination of:

```text
less chatter
fewer duplicate impulses
better transition commitment timing
better restraint after commitment
better handling of weak accumulated evidence
improved eyes-on musical coherence
```

without:

```text
suppressed legitimate repeated gestures
worse continuous behavior
fragile threshold tuning
meaningful runtime/export complexity
training instability disproportionate to the benefit
```

If a plain head plus cooldown/hysteresis performs equivalently, prefer the simpler mechanism and record ALIF as rejected.

## Relationship to #104

- #104 owns the causal trunk family and general hidden-state contract.
- This ticket owns only the optional sparse event-memory/commitment mechanism after `h_t`.
- A win here must not force #104 to adopt a full spiking architecture.
- A loss here must not count against a recurrent/SSM trunk whose continuous behavior remains strong.

## Relationship to #107

- #107 owns the semantic Controls v2 action surface.
- This ticket may gate an already-defined sparse Control; it may not invent a new Physics bypass.
- An impulse remains a bounded generalized impulse with ordinary metric-native semantics.
- No musical signal enters Physics directly.

## Relationship to #102/#131

- readiness/structure supervision may shape `h_t`;
- future-aware teacher values remain training-only;
- event gating is a runtime causal decision mechanism, not a teacher shortcut;
- auxiliary prediction accuracy is not acceptance evidence unless downstream behavior improves.

## Acceptance / resolution

Close this research ticket with:

- [ ] At least one genuinely sparse Controls-v2 event class is identified with a concrete chatter/timing hypothesis.
- [ ] Plain-head, simple hysteresis/cooldown, LIF, and ALIF variants are implemented or each receives an explicit justified exclusion.
- [ ] LIF/ALIF consume causal learned hidden state rather than raw/spike-encoded AudioActivation.
- [ ] Event firing gates an ordinary bounded analog payload rather than replacing all action magnitude/direction with spike coding.
- [ ] Canonical-tick leak/threshold/reset/adaptation equations and state-reset semantics are documented.
- [ ] Synthetic probes verify accumulator, leak, threshold, reset, refractory, and adaptation behavior deterministically.
- [ ] Future-aware SongFormer/readiness targets never enter runtime event state directly.
- [ ] Comparisons hold trunk/input/output contracts fixed closely enough to attribute the result to the commitment mechanism.
- [ ] Event chatter, duplicate-trigger rate, timing, premature/late commitment, and retrigger behavior are reported.
- [ ] Held-out eyes-on Player coherence is evaluated separately from auxiliary/teacher agreement.
- [ ] Runtime/export/state-width/training complexity is reported.
- [ ] LIF receives an explicit keep/drop verdict.
- [ ] ALIF receives an explicit keep/drop verdict.
- [ ] If simple cooldown/hysteresis is equivalent, the simpler mechanism wins.
- [ ] No result silently expands into a full spiking Player architecture.

## Non-goals

- Do not replace the #104 Player trunk with a spiking neural network.
- Do not spike-encode raw audio, CycleBank, or AudioActivation.
- Do not route all continuous Controls through event spikes.
- Do not feed SongFormer or other offline-teacher outputs into runtime LIF/ALIF state.
- Do not make transition readiness a handcrafted runtime feature merely to charge the integrator.
- Do not add direct music-aware gates to Mandelbrot Physics or Julia rendering.
- Do not invent new Controls solely because they are convenient to spike.
- Do not accept ALIF because it is biologically inspired or conceptually elegant; it must beat simpler behavioral baselines on held-out evidence.
